File: neetcode/blind75/trees/valid_binary_search_tree.py
# https://neetcode.io/problems/valid-binary-search-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def checkValidRange(curr, left_bound, right_bound):
            if not curr:
                return True
            elif (curr.val <= left_bound or curr.val >= right_bound):
                return False

            return checkValidRange(curr.left, left_bound, curr.val) and checkValidRange(curr.right, curr.val, right_bound)

        return checkValidRange(root, float('-inf'), float('inf'))
        
        

File: neetcode/neet150_JS_2/Binary-Tree/valid-binary-search-tree.js
// https://neetcode.io/problems/valid-binary-search-tree

/*
recursive dfs, 
    for each node check if the value is within the bounds set by its parents

- Time: O(n)
- Space: O(n)
*/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root) {
        if (root === null) {
            return true
        }

        return this.dfs(root, Number.NEGATIVE_INFINITY, Number.POSITIVE_INFINITY)
    }

    dfs(node, leftBound, rightBound) {
        if (node === null) {
            return true
        }

        if (node.val > leftBound && node.val < rightBound) {
            return this.dfs(node.left, leftBound, node.val) && this.dfs(node.right, node.val, rightBound)
        } else {
            return false
        }
    }
}
